React Nesting: Navigation Challenge (#144)
neuefische/web-exercises: sessions/react-nesting/navigation/README.md

### 1. Header component

Task: Outsource the `<header>` into a `Header` component.

- In the `/src` folder, create a subfolder called `components`. This is where all component files will live.
- In `/src/components`, create a file called `Header.js`.
  - Export a `Header` component which returns the same `<header></header>` element as in the `App.js` file.
  - Don't forget the `className` attribute.
  - The function should receive a `children` prop and render it between the opening and closing `<header>` tags.
- In the `App.js` file, import the `Header` component and replace the `<header></header>` tags with it.

🎉 Congratulations, you've just created a wrapping `Header` component using the `children` prop!

### 2. Navigation component

Let's do (almost) the same with the navigation.

Task: Outsource the `<nav>` into a `Navigation` component.

- Create `src/components/Navigation.js`.
- The `Navigation` should receive a `children` prop used in its return statement.
  - 💡 Note that the `nav` does not have a `className`, so there is no need to add this attribute.
- In the `App.js` file, import the `Navigation` component and replace the `<nav></nav>` tags with it.

> Understood the `children` prop now? Let's dive deeper!

Reviewer: sounds a bit too sassy 😄 

```suggestion
Awesome! You are a true `children` prop professional now. Let's dive deeper!
```

### 3. Link component

Task: Create a `Link` component to render `<a>` tags.

- Create a file for the `Link` component.
- The `Link` component should receive the props
  - `href` which is passed to the `href` attribute of the `<a>` tag and
  - `children` passed between the opening and closing `<a>` tags.

> - Don't forget the `className` attribute.

Reviewer: This can be interpreted in two ways:

1. The `className` is hardcoded within the component
2. The `className` is a prop

I was tempted to do two to replace all links in the app… this should be a little more clear

- In the `App.js` file, import the `Link` component and replace the `<a></a>` tags with it.
  - Make sure to pass the correct prop(s?).

🎉 Awesome, you've created a very flexible `Link` component! Want more? You'll get more!

### 4. Image component

Task: Create an `Image` component to render `<img>` tags.

> 💡 This will be a helper component to create a `Logo` and `Avatar` component.

- The `Image` component returns an `<img>` element with `className="round-image"` and

> - receives the props `src` and `alt` and passes them to the `src` and `alt` attributes.
> 

Reviewer: ```suggestion
- receives the props `src` and `alt` and passes them to the `src` and `alt` attributes.
- as we don't need children here, this component should not have a `children` prop
```

### 5. Logo and Avatar component

Task: Create a `Logo` and `Avatar` component, respectively.

> - Use the same JSX as in the `App.js` as basis (i.e. include the wrapping `<a>` or `<button>` tag), but

Reviewer: ```suggestion
- Use the same JSX as in the `App.js` as basis (copy the tags including `<a>` or `<button>` tag and their children)
```

  - import the `Image` component and use it.
  - make sure to import the correct `jpg` file.
- In the `App.js` file, import the `Logo` and `Avatar` component and replace the relevant JSX to use them.
- Check that the UI still looks the same.

> You should now have an `App` component returning only your custom components! Nice work 🎉

Reviewer: ````suggestion
You should now have an `App` component returning only your custom components (besides `main`)! Nice work 🎉

To check, you App.js file should look something like this now:

```js
import { Avatar } from "./components/Avatar";
import { Header } from "./components/Header";
import { Link } from "./components/Link";
import { Logo } from "./components/Logo";
import { Navigation } from "./components/Navigation";
import "./styles.css";

export default function App() {
  return (
    <>
      <Header>
        <Logo />
        <Navigation>
          <Link href="#home">Home</Link>
          <Link href="#about">About</Link>
          <Link href="#impressum">Impressum</Link>
        </Navigation>
        <Avatar />
      </Header>
      <main>content goes here…</main>
    </>
  );
}
```
````
